# Computational Theory - Shane Walsh - G00406694

## Task 1: Binary Representations
Create four functions in Python, demonstrating their use with examples and tests.
1. The function rotl(x, n=1) that rotates the bits in a 32-bit unsigned integer to the left n places.

2. The function rotr(x, n=1) that rotates the bits in a 32-bit unsigned integer to the right n places.

3. The function ch(x, y, z) that chooses the bits from y where x has bits set to 1 and bits in z where x has bits set to 0.

4. The function maj(x, y, z) which takes a majority vote of the bits in x, y, and z.

The output should have a 1 in bit position i where at least two of x, y, and z have 1's in position i.
All other output bit positions should be 0.

## 1. The function rotl(x, n=1)

This function goes through a process of rotating bits in a 32-bit unsigned integer to a direction. In this case, to the left by a a set number (such as 1 or 2) of positions. To implement this, we first need to understand the syntax for Bitwise shifting.

Bitwise left shift is <<
Bitwise right shift is >>

In [70]:
# Bitwise shifting of a number
x = 0b1100
print("Base value of X: " + bin(x)) # Base value
print("Shifted to the left by 2: " + bin(x << 2)) # Shift left by 2
print("Shifted to the right by 2: " + bin(x >> 2)) # Shift right by 2


Base value of X: 0b1100
Shifted to the left by 2: 0b110000
Shifted to the right by 2: 0b11


Now its important that to avoid losing any bits during the shift process, we can avoid this with what's known as masking. This ensures that our result doesn't grow larger by mistake and maintains the 32 bit size of my result.

In [71]:
# Masking examples here
x = 0b1100

The final step to tie this together is bitwise operations, which are your classic AND, OR, NOT and XOR gates. These allow us to control our results dynamically......

In [72]:
# Examples of bitwise operations
x = 0b1100
y = 0b1010
print(bin(x & y)) # Bitwise AND
print(bin(x | y)) # Bitwise OR
print(bin(x ^ y)) # Bitwise XOR
print(bin(~x)) # Bitwise NOT

0b1000
0b1110
0b110
-0b1101


In the end our rotation function looks like the following. It shifts bits left by n positions with (x << n). It then shifts bits that would be lost when shifting left, to the right with (x >> (32 - n)). We put an OR operator between these to combine them. Lastly, we use an AND operator to bring in our masking for ensuring to result stays at 32 bits. 

In [73]:
import numpy as np
import math

def rotl(x, n=1):
    return ((x << n) | (x >> (32 - n))) & 0xFFFFFFFF # << shifts the bits to left, >> shifts the bits to right.

x = 0b1100
print(f'rotl(0b1100, 2): {bin(rotl(x, 2))}')
print(f'rotl(0b1100, 3): {bin(rotl(x, 3))}')

rotl(0b1100, 2): 0b110000
rotl(0b1100, 3): 0b1100000


# Demonstrating the use of Rotl(x, n=1) 

## 2. The function rotr(x, n=1)

In [74]:
def rotr(x, n=1):
    return (x >> n) | ((x << (32 - n)) & 0xFFFFFFFF)

x = 0b1100

# Test the function.
print(f'rotr(0b1100, 2): {bin(rotr(x, 2))}')

rotr(0b1100, 2): 0b11


## 3. The function ch(x, y, z)

For this function we need to make use of bit selection logic using bitwise operators on our chosen binary values. In this particular implementation: 

- For each bit position of x being 1, we choose the associated or corresponding bit from y.
- For each bit position of x being 0, we choose the corresponding bit from z.

Let me further break down how to accomplish this: 
Using (x & y) will take bits from y where x has 1s.
Using  ()

In [75]:
def ch(x, y, z): # choose between x, y, z. 
    return (x & y) ^ (~x & z) # & is bitwise AND, ^ is bitwise XOR, ~ is bitwise NOT. So this returns x AND y XOR NOT x AND z.

x = 0b1100
y = 0b1010
z = 0b1001

print(f'ch(0b1100, 0b1010, 0b1001): {bin(ch(x, y, z))}')

ch(0b1100, 0b1010, 0b1001): 0b1001


## 4. The function maj(x, y, z)

In [76]:
def maj(x, y, z): # Takes a majority vote of x, y, z.
    return (x & y) ^ (x & z) ^ (y & z) # ^ means XOR. So this returns x AND y XOR x AND z XOR y AND z.

x = 0b1100
y = 0b1010
z = 0b1001

x1 = 0b1100
y1 = 0b1010
z1 = 0b1010

# Examples
print(f'maj(0b1100, 0b1010, 0b1001): {bin(maj(x, y, z))}')
print(f'maj(0b1100, 0b1010, 0b1010): {bin(maj(x1, y1, z1))}')

maj(0b1100, 0b1010, 0b1001): 0b1000
maj(0b1100, 0b1010, 0b1010): 0b1010


## Task 2: Hash Functions
The following hash function is from The C Programming Language by Brian Kernighan and Dennis Ritchie.
Convert it to Python, test it, and suggest why the values 31 and 101 are used.

unsigned hash(char *s) {
    unsigned hashval;
    for (hashval = 0; *s != '\0'; s++)
        hashval = *s + 31 * hashval;
    return hashval % 101;
}

I'll now convert this hash function from the C Programming language to python. This will require quite a few changes to the code. We don't need to declare variable types. While the hash function in C uses pointer's and null terminator checking, Python makes this much simpler as we can just iterate directly over the string. The Hash function that we're converting in this process is from page 128 - 129 of [The C Programming Language](https://seriouscomputerist.atariverse.com/media/pdf/book/C%20Programming%20Language%20-%202nd%20Edition%20(OCR).pdf) by Brian Kernighan and Dennis Ritchie.

In [77]:
def Hash_string(s): # No need to declare var, just use it with the input, no pointer.
    hashval = 0 # No use of 'unsigned' in Python.
    # Iterate directly over string.
    for c in s:
        hashval = ord(c) + 31 * hashval # ord() returns ASCII value of char, 31 is prime number, good for hashing.
    return hashval % 101 # Modulus is unchanged, 101 is prime number, reduces collisions.

The final conversion would look like this. Rather than dereferencing the character pointer in C, we can use Python's [ord() function](https://docs.python.org/3.4/library/functions.html#ord) to pull out the relevant [ASCII value](https://www.ascii-code.com/) of a character. THe modulus element on the hashVal has no necessary changes.

The numbers 31 and 101 are prime numbers. To my knowledge, using prime numbers for hash values gives us a smoother distribution. 101 helps determine our hash table size and reduce collisions which can quite easily come up. 

In [78]:
# Testing Hash function
print(f'Hash_string("hello"): {Hash_string("hello")}')
print(f'Hash_string("world"): {Hash_string("world")}')
print(f'Hash_string("hello world"): {Hash_string("hello world")}')

Hash_string("hello"): 17
Hash_string("world"): 34
Hash_string("hello world"): 13


## Task 3: SHA256

Write a Python function that calculates the SHA256 padding for a given file.  
The function should take a file path as input.  
It should print, in hex, the padding that would be applied to it.  
The [specification](https://doi.org/10.6028/NIST.FIPS.180-4) states that the following should be appended to a message:  

- a`1` bit;
- enough `0` bits so the length in bits of padded message is the smallest possible multiple of 512;
- the length in bits of the original input as a big-endian 64-bit unsigned integer.

The example in the specification is a file containing the three bytes `abc`:  

```python
01100001 01100010 01100011
```

The output would be:  

```python
80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 18
```

So for this task, we need to calculate the SHA256 padding for a file supplied in the method as input. Ideally, I should adhere to the NIST FIPS 180-4 specification during this process. --- Add Ref here. In the end I'll be returning the padding as a collection of bytes.

We first need to read in the file as a collection of bytes. This is farely easy and selfexplanatory to implement. Right after that we calculate the original content length in bits. 

In [79]:

# Create a file called file.txt with our three bytes in it for example purposes.
file = 'file.txt'

with open(file, 'w') as f: # For testing purposes, write some text to a file.
    f.write('abc')

# Read in the file
with open(file, 'rb') as f:
    data = f.read()

# Calculate length in bits of the file
length = len(data) * 8

print(f'Length of file in bits: {length}')

Length of file in bits: 24


Following the length calculation, we should follow the SHA-256 specification for applying padding. First off, I'll add a single '1' bit as a hexidecimal (0x80). We need to add enough '0' bits so the total length of our file (including both the original length, the padding and 64 bits) is a multiple of 512 bits. 

In [80]:
### Example Code here


We use the calculation padding_needed = 56 - ((len(data) + 1) % 64) to ensure we leave exactly 64 bits (or 8 bytes) at the end of the file. 

In [81]:
### Example Code here

Following this, we add back the original length of the file as a 64 bit big-endian unsigned integer. 

In [82]:
### Final Sha256 Method

# Calculate the SHA256 padding of a string from a giving file.
def SHA256_padding(file):
    # Read in the file
    with open(file, 'rb') as f:
        data = f.read()

    # Calculate length in bits of the file, multiply by 8 to get bits as there is 8 bits in a byte.
    original_length = len(data) * 8

    # Add a '1' bit to the end of the data, in hex this would be 0x80 or 10000000 in python.
    padding = bytearray([0x80])

    # Add enough '0' bits so the total length is 448 mod 512 bits.
    # So the original length + 1 + padding) % 64 = 56, we need 56 to be the remainder. 
    padding_zeros = 56 - ((len(data) + 1) % 64)
    # if we reach the 56-bye mark, add another block. So we dynamically add padding needed to the file. 
    if padding_zeros < 0:
        padding_zeros += 64.

    # Add original length of file in bits, this must be a 64 big-endian unsigned integer. So as to not overflow.
    padding.extend([0] * padding_zeros)
    padding.extend(original_length.to_bytes(8, byteorder='big'))


    # Print this padding in hexidecimal. 
    print("SHA256 padding in hexidecimal format: ")
    for i, byte in enumerate(padding):
        print(f'{byte:02x}', end=' ')
        if (i + 1) % 25 == 0: # Print 25 bytes per line for simplicity and readability.
            print()
        if len(padding) % 25 != 0:
            print() # Print a newline if the padding is not a multiple of 25 bytes.
    
    return padding # Return the padding at the end. 

## Example explanation

In [83]:
### Example usage with a test file. 
def test_sha256_padding():
    # Create a test file with "abc"
    with open("test_file.txt", "wb") as f:
        f.write(b"abc")
    
    # Calculate and print the padding
    padding = SHA256_padding("test_file.txt")
    return padding

# Test the function
padding = test_sha256_padding()


SHA256 padding in hexidecimal format: 
80 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 

00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
00 

00 
00 
00 
00 
00 
00 
00 
00 
00 
00 
18 


## Task 4: Prime Numbers

Calculate the first 100 prime numbers using two different algorithms.  
Any algorithms that are well-established and works correctly are okay to use.  
Explain how the algorithms work.

Insert links here for each algorithm

## Division with square roots
The basic process for this algorithm for checking if a number is prime involves finding out if its divisible by any smaller numbers than it. We do this through using its square roots. This is useful, as we only need to check if a number divides into our square root of n because: 
- If its a multiple of n, then either a or b must be included in its square root. 




In [84]:
# Insert edge cases here

We than proceed to check each number in sequence, adding it to our list if its a prime and leaving it be if it isn't one. We continue this process until we hopefully reach 100 prime numbers. 

In [85]:
# Insert main while loop here

In [86]:
import numpy as np
import matplotlib.pyplot as plt
import time

def test_prime(n):
    # Test if a number is prime by checking if it is divisible by any number up to its square root. n is the number to test.

    # We only need to test numbers of the form "6k±1" since all primes higher than a 3 are of this form. 
    # In order to do this, we must check the edge cases of 2 and 3 separately. Once we've checked 2 we know its not even.
    if n <= 1: # Check if n is less than or equal to 1, if so, return False.
        return False
    if n <= 3: # Check if n is less than or equal to 3, if so, return True.
        return True
    if n % 2 == 0 or n % 3 == 0: # Check if n is divisible by 2 or 3, if so, return False.
        return False
    
    # Check all numbers of the form "6k±1" up to the square root of n.
    i = 5 # Start at 5, since we've already checked 2 and 3.
    while i * i <= n: # Check if i squared is less than or equal to n.
        if n % i == 0 or n % (i + 2) == 0: # Check if n can be divided by i or i + 2, return False if so.
            return False
        i += 6 # Increment i by 6, since we're checking numbers of the "6k±1" form.
    return True

    
    

Here is a simple function for running that algorithm a number of times equal to the user's desire.

In [87]:
def find_primes(count):
    # Find the first x number of primes based on the count given.
    primes = [] # Create an empty list to store our found primes.
    num = 2 # Start at 2, the first prime number.

    # Loop until we have found the desired number of primes.
    while len(primes) < count:
        if test_prime(num):
            primes.append(num)
        num += 1

    return primes

# Testing with Examples

In [88]:
primes = find_primes(100)

print(primes)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541]


## Sieve Algorithm
THe previous method works overall, but it's quite a brute force approach to the issue. Instead we could explore using the Sieve approach (reference links here). Instead of testing every individual number as we progress through sequentially, we can eliminate the composite numbers systematically, allowing us to widdle it down to the 100 primes we need. 

We start with a list of numbers, we'll choose a minimum and a maximum of 2 and 1000.

At first, we'll assume all numbers are primes. Then as we loop through them all, we'll mark all the multiples of each prime as false or not prime. Any numbers that aren't marked by this are found as primes. This approach is quite a bit more effective and efficient in comparison to our last algorithm. We only need to search up to a certain limit, because larger factors of a prime would have been marked by their smaller counterparts already. Additionally, each number is marked only once for each of its prime numbers, rather than repeatedly parsing over familiar numbers within the previous algorithm.

In [89]:
def findPrimesUsingSieve(count):
    # Find the first x number of primes based on the count given.
    limit = 1000 # Set a limit for the sieve.

    # Initialize the sieve list.
    sieve = [True] * (limit + 1) # Create a list of True values, one for each number up to the limit.
    # Set the first two values to False, since they are not prime.
    sieve[0] = False 
    sieve[1] = False

    # Loop through the sieve and set all multiples of each number to False.
    for i in range(2, int(np.sqrt(limit)) + 1): # Loop through up to square root of limit.
        if sieve[i]: # If the number is prime.
            for j in range(i * i, limit + 1, i): # Loop through all multiples of the number.
                sieve[j] = False # Set the multiple to False.

    # Loop through the sieve and add all prime numbers to the list.
    primes = [] # Create empty list to store found primes.
    for i in range(limit + 1):
        if sieve[i]: # If number is prime.
            primes.append(i) # Add number to list.
            if len(primes) == count: # Once we reach count, break loop.
                break

    return primes


print(findPrimesUsingSieve(100))

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541]


In [90]:
# Compare the results to ensure they match
print(f"\nBoth algorithms produce the same results: {find_primes(100) == findPrimesUsingSieve(100)}")


Both algorithms produce the same results: True


Its worth noting that the Sieve algorithm seems far more effective, particularly in terms of speed. This is generally due to the fact that it eliminates multiple composite numbers with each of its loops, instead of individually. It marks off all the multiples of the known primes, avoiding redundancy. As you can imagine, this is especially effective for larger numbers that possess many factors. Sieve has a handy amorphous nature to it, working across the entire range of numbers rather than singularly. 

### Task 5: Roots
Calculate the first 32 bits of the fractional part of the square roots of the first 100 prime numbers.

So our aim with this task is to take the first 100 prime numbers (which we've gathered in our previous task) and calculate the first 32 bits of eah of their fractional parts of their square roots. First off, lets get the square root of each prime number. To do this in python, you multiply against the power of 0.5 of your prime. As shown below, for example purposes I've used 7 as its a fairly common prime number. 

In [91]:
prime = 7
root = prime ** 0.5 # Square root is written as a power of 0.5 in Python.
print(f'root: {root}')

root: 2.6457513110645907


After this we need to extract the fractional parts of our square root. The fractional parts is (as the name suggests) anything after the decimal point. So for the square root of 7: 2.6457513110645907, it'd be 6457513110645907. To extract the fractional part of our square rooted prime in python, I think my best option is to use math.floor and take that away from our root to get the fractional part. 

In [92]:
frac = root - math.floor(root) # Subtract the floor of the root from the root to get the fractional part.
print(f'frac: {frac}')

frac: 0.6457513110645907


Next up is where we dive into bits territory. In this case, we need to represent our fractional part (such as .6457513110645907) as a 32 bit value (like ). We can do this by 

In [ ]:
frac = frac * (2 ** 32) # Multiply the fractional part by 2^32 to get the fractional part in integer form.
print(f'frac: {frac}')

# Convert to integer through truncation so we only have the integer part.
bits = int(frac)
print(f'bits: {bits}')

frac: 2773480762.37154
bits: 2773480762


Now that we've accomplished this algorithm step by step. I'll combine this together into a singular method.